source/common/http/filter_manager.cc

```diff
@@ -183,11 +183,15 @@ bool ActiveStreamFilterBase::commonHandleAfterDataCallback(FilterDataStatus stat
       buffer_was_streaming = status == FilterDataStatus::StopIterationAndWatermark;
       commonHandleBufferData(provided_data);
     } else if (complete() && !hasTrailers() && !bufferedData()) {
-      // If this filter is doing StopIterationNoBuffer and this stream is terminated with a zero
-      // byte data frame, we need to create an empty buffer to make sure that when commonContinue
-      // is called, the pipeline resumes with an empty data frame with end_stream = true
-      ASSERT(end_stream_);
-      bufferedData() = createBuffer();
+      if (end_stream_) {
+        // If this filter is doing StopIterationNoBuffer and this stream is terminated with a zero
+        // byte data frame, we need to create an empty buffer to make sure that when commonContinue
+        // is called, the pipeline resumes with an empty data frame with end_stream = true
+        bufferedData() = createBuffer();
+      } else {
+        // Otherwise the entire connection was closed.
+        ASSERT(parent_.state_.local_complete_ || parent_.state_.remote_complete_);
```

Comment:
I'm not sure if the assertion verifies that "the entire connection was closed": these flags are set when the filter sees a encode or decode end_stream=true, which aren't related to the connection lifetime afaik.

Would it be simpler to just set end_stream_ = true for the filter that triggers the invalid local reply?

Reply:
Perhaps I'm getting my terminology confused then. I see 3 terms used in HCM: "end stream, reset stream, reset connection". 

I followed [this code path](https://github.com/envoyproxy/envoy/blob/3f740bc90d32ac442e70da46ebf4d03a1bdb479d/source/common/http/filter_manager.cc#L850) from `sendLocalReply` to determine that these flags would be set. These flags are set in `resetStream()` [here](https://github.com/envoyproxy/envoy/blob/3f740bc90d32ac442e70da46ebf4d03a1bdb479d/source/common/http/conn_manager_impl.cc#L181).

Would you suggest setting `end_stream_ = true` somewhere in this code path? I do not understand why `resetStream()` does not set `end_stream_ = true`.

Reply:
Looking at the code I think we're just resetting the stream (which is connection close for HTTP/1.1, but not HTTP/2). I think `remote_complete_` would be irrelevant here, so if anything we'd just verify that `ASSERT(parent_.state_.local_complete_)` is true as a proxy for checking whether we've reset the stream. 

I think one option would be to set `end_stream_` to true in `ActiveStreamEncoderFilter::sendLocalReply` after having issued the local reply, though that might be problematic if the callback that issued the filter callback has end_stream=true. 

Overall I'm not sure what the value of these new asserts are vs just removing the existing assert altogether. Maybe we should have a new flag that indicates reset stream that we can assert on? @alyssawilk might have thoughts

Reply:
I don't think we have a flag like this, we'd probably want to set it in maybeEndEncode when end_stream=true. We do something like this for the decoding path: remote_complete_ basically means "the FM has received the full request". 

I think the names of local vs remote complete flags is a bit confusing, as I've definitely been confused about whether we're talking about local encoding vs decoding or whether we're talking about the remote upstream vs downstream. We can probably leave that for a different PR though.

Reply:
Ok, I understand in what conditions the buffer should be created. I implemented it as you described: Setting `end_stream_encoded_` in `maybeEndEncode`.

I am still struggling to understand the need for the empty buffer. I cannot find any test cases that break when I remove this functionality, but perhaps I'm missing something. Would appreciate if you could double check.